src/sdam/srv_polling/mod.rs

```diff
@@ -0,0 +1,140 @@
+
+pub(crate) struct SrvPollingMonitor {
+    initial_hostname: String,
+    resolver: Option<SrvResolver>,
+    topology: WeakTopology,
+    rescan_interval: Option<Duration>,
+    client_options: ClientOptions,
+}
+
+impl SrvPollingMonitor {
+    pub(crate) fn new(topology: WeakTopology) -> Option<Self> {
+        let client_options = topology.client_options().clone();
+
+        let initial_hostname = match client_options.original_srv_hostname() {
+            Some(hostname) => hostname.clone(),
+            None => return None,
+        };
+
+        Some(Self {
+            initial_hostname,
+            resolver: None,
+            topology,
+            rescan_interval: None,
+            client_options,
+        })
+    }
+
+    /// Starts a monitoring task that periodically performs SRV record lookups to determine if the
+    /// set of mongos in the cluster have changed. A weak reference is used to ensure that the
+    /// monitoring task doesn't keep the topology alive after the client has been dropped.
+    pub(super) fn start(topology: WeakTopology) {
+        RUNTIME.execute(async move {
+            if let Some(mut monitor) = Self::new(topology) {
+                monitor.execute().await;
+            }
+        });
+    }
+
+    async fn execute(&mut self) {
+        while let Some(topology) = self.topology.upgrade() {
+            let state = topology.clone_state().await;
+
+            if state.is_sharded() || state.is_unknown() {
+                let hosts = self.lookup_hosts().await;
+                self.update_hosts(hosts, topology, state).await;
+            }
+
+            RUNTIME
+                .delay_for(self.rescan_interval.unwrap_or(DEFAULT_RESCAN_SRV_INTERVAL))
+                .await;
+        }
+    }
+
+    pub(crate) async fn update_hosts(
+        &mut self,
+        hosts: Result<Vec<StreamAddress>>,
+        topology: Topology,
+        mut topology_state: TopologyState,
+    ) {
+        let hosts = match hosts {
+            Ok(hosts) if hosts.is_empty() => {
+                self.no_valid_hosts(None);
+
+                return;
+            }
+            Ok(hosts) => hosts,
+            Err(err) => {
+                self.no_valid_hosts(Some(err));
+
+                return;
+            }
+        };
+
+        // TODO: Use TTL values to determine delay duration once `trust-dns-resolver`
+        // releases again.
+        self.rescan_interval = None;
+
+        let diff = topology_state.update_hosts(&hosts.into_iter().collect(), &self.client_options);
+        topology.update_state(diff, topology_state).await;
+    }
+
+    async fn lookup_hosts(&mut self) -> Result<Vec<StreamAddress>> {
+        let initial_hostname = self.initial_hostname.clone();
+        let resolver = self.get_or_create_srv_resolver().await?;
+        let mut new_hosts = Vec::new();
+
+        for host in resolver.get_srv_hosts(&initial_hostname).await? {
+            #[allow(clippy::single_match)]
```

Comment:
Although we could just use an `if let` here, I figured we should just do the `match` now so that the TODO could be in a more correct place.